File: bada_ruby/lib/bada/qc/isa.rb
# frozen_string_literal: true

module Bada
  module QC
    # ISA — the pseudo quantum computer's instruction set (BadaQASM) and its
    # von-Neumann encoding into disk words.
    #
    # Each instruction is a fixed 4-word record:  [opcode, a, b, theta]
    # so the *program* lives in the same disk memory as the *data* (state
    # vector) — a stored-program (von Neumann) machine.
    module ISA
      module_function

      RECORD_WORDS = 4

      # opcode table (4-bit, decoded by the simulated semiconductor decoder)
      OPCODES = {
        "NOP"     => 0,
        "H"       => 1,  # Hadamard on a
        "X"       => 2,  # Pauli-X on a
        "Y"       => 3,  # Pauli-Y on a
        "Z"       => 4,  # Pauli-Z on a
        "S"       => 5,  # phase S on a
        "T"       => 6,  # phase T on a
        "CX"      => 7,  # CNOT control a, target b
        "RX"      => 8,  # rotation about X by theta on a
        "RZ"      => 9,  # rotation about Z by theta on a
        "MEASURE" => 10, # collapse a (records classical bit)
        "HALT"    => 11
      }.freeze

      OPCODE_BITS = 4
      NAMES = OPCODES.invert.freeze

      # Parse BadaQASM source into an array of [opcode, a, b, theta] records.
      #
      #   # comment
      #   H 0
      #   CX 0 1
      #   RZ 1 0.7853981634
      #   MEASURE 0
      #   HALT
      def assemble(source)
        program = []
        source.each_line do |raw|
          line = raw.strip.sub(/#.*\z/, "").strip
          next if line.empty?
          tok = line.split(/\s+/)
          mnem = tok[0].upcase
          op = OPCODES.fetch(mnem) { raise "unknown instruction: #{mnem}" }
          a = tok[1] ? tok[1].to_i : 0
          b = tok[2] ? tok[2].to_i : 0
          theta =
            case mnem
            when "RX", "RZ" then (tok[2] || tok[1]).to_f # theta is last arg
            else 0.0
            end
          # For RX/RZ the form is "RX <qubit> <theta>", so b holds nothing useful:
          if %w[RX RZ].include?(mnem)
            a = tok[1].to_i
            theta = tok[2].to_f
            b = 0
          end
          program << [op, a, b, theta]
        end
        program
      end

      def mnemonic(opcode)
        NAMES[opcode] || "?"
      end
    end
  end
end

File: apps/core/src/main/java/bada/qc/Isa.java
package bada.qc;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * ISA — BadaQASM instruction set and its von-Neumann encoding into disk words.
 * Each instruction is a 4-word record [opcode, a, b, theta]. Java port of
 * Bada::QC::ISA.
 */
public final class Isa {
    public static final int RECORD_WORDS = 4;
    public static final int OPCODE_BITS = 4;

    public static final Map<String, Integer> OPCODES = new LinkedHashMap<>();
    public static final Map<Integer, String> NAMES = new LinkedHashMap<>();
    static {
        String[] mn = {"NOP", "H", "X", "Y", "Z", "S", "T", "CX", "RX", "RZ", "MEASURE", "HALT"};
        for (int i = 0; i < mn.length; i++) {
            OPCODES.put(mn[i], i);
            NAMES.put(i, mn[i]);
        }
    }

    /** An assembled instruction record. */
    public static final class Instr {
        public final int op, a, b;
        public final double theta;
        public Instr(int op, int a, int b, double theta) {
            this.op = op; this.a = a; this.b = b; this.theta = theta;
        }
    }

    /** Parse BadaQASM source into records. */
    public static List<Instr> assemble(String source) {
        List<Instr> program = new ArrayList<>();
        for (String raw : source.split("\n")) {
            String line = raw.replaceAll("#.*$", "").trim();
            if (line.isEmpty()) continue;
            String[] tok = line.split("\\s+");
            String mnem = tok[0].toUpperCase();
            Integer op = OPCODES.get(mnem);
            if (op == null) throw new IllegalArgumentException("unknown instruction: " + mnem);
            int a = tok.length > 1 ? (int) parseNum(tok[1]) : 0;
            int b = tok.length > 2 ? (int) parseNum(tok[2]) : 0;
            double theta = 0.0;
            if (mnem.equals("RX") || mnem.equals("RZ")) {
                a = tok.length > 1 ? (int) parseNum(tok[1]) : 0;
                theta = tok.length > 2 ? parseNum(tok[2]) : 0.0;
                b = 0;
            }
            program.add(new Instr(op, a, b, theta));
        }
        return program;
    }

    private static double parseNum(String s) {
        try { return Double.parseDouble(s); } catch (NumberFormatException e) { return 0.0; }
    }

    public static String mnemonic(int opcode) {
        return NAMES.getOrDefault(opcode, "?");
    }

    private Isa() { }
}
